pkg/ottl/boolean_value.go

```diff
@@ -8,195 +8,280 @@ import (
 	"fmt"
 )
 
-// boolExpressionEvaluator is a function that returns the result.
-type boolExpressionEvaluator[K any] func(ctx context.Context, tCtx K) (bool, error)
-
 // BoolExpr represents a condition in OTTL
-type BoolExpr[K any] struct {
-	boolExpressionEvaluator[K]
+type BoolExpr[K any] interface {
+	Eval(ctx context.Context, tCtx K) (bool, error)
+
+	unexported()
 }
 
-// Eval evaluates an OTTL condition
-func (e BoolExpr[K]) Eval(ctx context.Context, tCtx K) (bool, error) {
-	return e.boolExpressionEvaluator(ctx, tCtx)
+type literalBoolExpr[K any] = literalExpr[K, bool]
+
+func newAlwaysTrue[K any]() BoolExpr[K] {
+	return &literalBoolExpr[K]{val: true}
 }
 
-//nolint:unparam
-func not[K any](original BoolExpr[K]) (BoolExpr[K], error) {
-	return BoolExpr[K]{func(ctx context.Context, tCtx K) (bool, error) {
-		result, err := original.Eval(ctx, tCtx)
-		return !result, err
-	}}, nil
+func newAlwaysFalse[K any]() BoolExpr[K] {
+	return &literalBoolExpr[K]{val: false}
 }
 
-func alwaysTrue[K any](context.Context, K) (bool, error) {
-	return true, nil
+func newNot[K any](expr BoolExpr[K]) BoolExpr[K] {
+	if f, ok := expr.(*literalBoolExpr[K]); ok {
+		return &literalBoolExpr[K]{val: !f.val}
+	}
+
+	return &notBoolExpr[K]{expr: expr}
 }
 
-func alwaysFalse[K any](context.Context, K) (bool, error) {
-	return false, nil
+type notBoolExpr[K any] struct {
+	expr BoolExpr[K]
+}
+
+func (*notBoolExpr[K]) unexported() {}
+
+// Eval evaluates an OTTL condition
+func (e *notBoolExpr[K]) Eval(ctx context.Context, tCtx K) (bool, error) {
+	val, err := e.expr.Eval(ctx, tCtx)
+	return !val, err
 }
 
-// builds a function that returns a short-circuited result of ANDing
+// newAndExprs a BoolExpr that returns a short-circuited result of ANDing
 // boolExpressionEvaluator funcs
-func andFuncs[K any](funcs []BoolExpr[K]) BoolExpr[K] {
-	return BoolExpr[K]{func(ctx context.Context, tCtx K) (bool, error) {
-		for _, f := range funcs {
-			result, err := f.Eval(ctx, tCtx)
-			if err != nil {
-				return false, err
-			}
-			if !result {
-				return false, nil
+func newAndExprs[K any](exprs []BoolExpr[K]) BoolExpr[K] {
+	newExprs := make([]BoolExpr[K], 0, len(exprs))
+	for i := range exprs {
+		// Check if any literal function that returns false then we can return false.
+		// If returns true just skip evaluation.
+		if f, ok := exprs[i].(*literalBoolExpr[K]); ok {
+			if f.getValue() == false {
+				return newAlwaysFalse[K]()
 			}
+			continue
 		}
-		return true, nil
-	}}
+		newExprs = append(newExprs, exprs[i])
+	}
+
```

Comment:
Would we've any gain doing something like this when all exprs are literals and evaluates to `true`?

```go
	if len(newExprs) == 0 {
		return newAlwaysTrue[K]()
	}
```